File: src/relations.ts
import {
  RelationSpeciesArchetype,
  RelationArchetypePower,
  RelationBiomeStructure,
  RelationVibePalette,
  WeightedEntry
} from './types.js';

export interface RelationData {
  species_archetype: { relations: RelationSpeciesArchetype[] };
  archetype_power_source: { relations: RelationArchetypePower[] };
  biome_structure: { relations: RelationBiomeStructure[] };
  vibe_palette: { relations: RelationVibePalette[] };
  conflicts: { conflicts: any[] };
}

export function applySpeciesArchetype(
  chosenSpecies: string | undefined,
  archetypeCandidates: WeightedEntry[],
  rel: RelationData['species_archetype']
): string[] {
  const applied: string[] = [];
  if (!chosenSpecies) return applied;
  for (const r of rel.relations) {
    if (r.species === chosenSpecies) {
      const match = archetypeCandidates.find(a => a.token === r.archetype);
      if (match) {
        match.weight = Math.max(1, Math.round(match.weight * r.weight_multiplier));
        applied.push(`species_archetype:${chosenSpecies}->${r.archetype} x${r.weight_multiplier}`);
      }
    }
  }
  return applied;
}

export function applyArchetypePower(
  chosenArchetype: string | undefined,
  powerCandidates: WeightedEntry[],
  rel: RelationData['archetype_power_source']
): string[] {
  const applied: string[] = [];
  if (!chosenArchetype) return applied;
  for (const r of rel.relations) {
    if (r.archetype === chosenArchetype) {
      const match = powerCandidates.find(p => p.token === r.power_source);
      if (match) {
        match.weight = Math.max(1, Math.round(match.weight * r.weight_multiplier));
        applied.push(`archetype_power:${chosenArchetype}->${r.power_source} x${r.weight_multiplier}`);
      }
    }
  }
  return applied;
}

export function applyBiomeStructure(
  chosenBiome: string | undefined,
  structureCandidates: WeightedEntry[],
  rel: RelationData['biome_structure']
): string[] {
  const applied: string[] = [];
  if (!chosenBiome) return applied;
  for (const r of rel.relations) {
    if (r.biome === chosenBiome) {
      const match = structureCandidates.find(s => s.token === r.structure);
      if (match) {
        match.weight = Math.max(1, Math.round(match.weight * r.weight_multiplier));
        applied.push(`biome_structure:${chosenBiome}->${r.structure} x${r.weight_multiplier}`);
      }
    }
  }
  return applied;
}

export function applyVibePalette(
  chosenVibe: string,
  paletteCandidates: WeightedEntry[],
  rel: RelationData['vibe_palette']
): string[] {
  const applied: string[] = [];
  for (const r of rel.relations) {
    if (r.vibe === chosenVibe) {
      const match = paletteCandidates.find(p => p.token === r.palette);
      if (match) {
        match.weight = Math.max(1, Math.round(match.weight * r.weight_multiplier));
        applied.push(`vibe_palette:${chosenVibe}->${r.palette} x${r.weight_multiplier}`);
      }
    }
  }
  return applied;
}

File: engine/relations.py
"""
Relational weight adjustments between different dimensions
"""
from typing import List, Dict, Any
import copy

def apply_species_archetype(entries: List[Dict[str, Any]], context: Dict[str, Any], 
                           relations: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Apply species-archetype relational weighting"""
    species_tokens = context.get('tokens', {}).get('species', [])
    if not species_tokens:
        return entries
    
    relations_data = relations.get('species_archetype', {}).get('relations', [])
    if not relations_data:
        return entries
    
    result = []
    for entry in entries:
        entry_copy = copy.deepcopy(entry)
        
        for species in species_tokens:
            for relation in relations_data:
                if (relation.get('species') == species and
                    relation.get('archetype') == entry_copy.get('token')):
                    multiplier = relation.get('weight_multiplier', 1.0)
                    entry_copy['weight'] = int(entry_copy.get('weight', 1) * multiplier)
                    context.setdefault('relations_applied', []).append(
                        f"species-archetype: {species} -> {entry_copy.get('token')} ({multiplier}x)"
                    )
                    break
        
        result.append(entry_copy)
    
    return result

def apply_archetype_power(entries: List[Dict[str, Any]], context: Dict[str, Any],
                         relations: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Apply archetype-power_source relational weighting"""
    archetype_tokens = context.get('tokens', {}).get('archetypes', [])
    if not archetype_tokens:
        return entries
    
    relations_data = relations.get('archetype_power_source', {}).get('relations', [])
    if not relations_data:
        return entries
    
    result = []
    for entry in entries:
        entry_copy = copy.deepcopy(entry)
        
        for archetype in archetype_tokens:
            for relation in relations_data:
                if (relation.get('archetype') == archetype and
                    relation.get('power_source') == entry_copy.get('token')):
                    multiplier = relation.get('weight_multiplier', 1.0)
                    entry_copy['weight'] = int(entry_copy.get('weight', 1) * multiplier)
                    context.setdefault('relations_applied', []).append(
                        f"archetype-power: {archetype} -> {entry_copy.get('token')} ({multiplier}x)"
                    )
                    break
        
        result.append(entry_copy)
    
    return result

def apply_biome_structure(entries: List[Dict[str, Any]], context: Dict[str, Any],
                         relations: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Apply biome-structure relational weighting"""
    biome_tokens = context.get('tokens', {}).get('biomes', [])
    if not biome_tokens:
        return entries
    
    relations_data = relations.get('biome_structure', {}).get('relations', [])
    if not relations_data:
        return entries
    
    result = []
    for entry in entries:
        entry_copy = copy.deepcopy(entry)
        
        for biome in biome_tokens:
            for relation in relations_data:
                if (relation.get('biome') == biome and
                    relation.get('structure') == entry_copy.get('token')):
                    multiplier = relation.get('weight_multiplier', 1.0)
                    entry_copy['weight'] = int(entry_copy.get('weight', 1) * multiplier)
                    context.setdefault('relations_applied', []).append(
                        f"biome-structure: {biome} -> {entry_copy.get('token')} ({multiplier}x)"
                    )
                    break
        
        result.append(entry_copy)
    
    return result

def apply_vibe_palette(entries: List[Dict[str, Any]], context: Dict[str, Any],
                      relations: Dict[str, Any]) -> List[Dict[str, Any]]:
    """Apply vibe-palette relational weighting"""
    vibe = context.get('vibe')
    if not vibe:
        return entries
    
    relations_data = relations.get('vibe_palette', {}).get('relations', [])
    if not relations_data:
        return entries
    
    result = []
    for entry in entries:
        entry_copy = copy.deepcopy(entry)
        
        for relation in relations_data:
            if (relation.get('vibe') == vibe and
                relation.get('palette') == entry_copy.get('token')):
                multiplier = relation.get('weight_multiplier', 1.0)
                entry_copy['weight'] = int(entry_copy.get('weight', 1) * multiplier)
                context.setdefault('relations_applied', []).append(
                    f"vibe-palette: {vibe} -> {entry_copy.get('token')} ({multiplier}x)"
                )
                break
        
        result.append(entry_copy)
    
    return result